bittensor/utils/tokenizer_utils.py

```diff
@@ -0,0 +1,910 @@
+    translation_max_len = counts.shape[0]  # maximum possible many-to-one length available in translation map
+
+    if many_len <= translation_max_len:
+        probs_from_copy /= counts[:many_len, :]  # divide probability mass by amount of paths crossing each token
+    else:  # limit probs_from token depth to max_len
+        probs_from_copy[:translation_max_len, :] /= counts
+
+    # === Reverse map std token to source sequences, gather avg. sequence prob ===
+    for map_len in translation_map['lengths'].keys():  # mutually exclusive over std tokens
+        from_idx = translation_map['lengths'][map_len]['from']  # [subset_size_std] one std token
+        to_idx = translation_map['lengths'][map_len]['to'].T  # [map_len, subset_size_std] many server token seq
+        if many_len < map_len:  # sequence beyond segment_count has min probability 0
+            to_idx = to_idx[:many_len, :]  # [segment_count, subset_size_std]
+        server_seq_tokens = probs_from_copy.gather(1, to_idx)  # [map_len, subset_size_std] gather sequences
+        probs_to[from_idx] = server_seq_tokens.sum(dim=0) / map_len  # [subset_size_std] in-place average approx.
+
+
+def translate_tokenizer_probs(probs: torch.FloatTensor, probs_std: torch.FloatTensor,
+                              offset_mapping: List[tuple], offset_mapping_std: List[tuple],
+                              tokenizer: PreTrainedTokenizerBase, std_tokenizer: PreTrainedTokenizerBase,
+                              split_map_cache: Dict[tuple, List[Dict[str, torch.Tensor]]],
+                              to_translation_map: Dict[str, Any], from_translation_map: Dict[str, Any],
+                              tokens: torch.LongTensor, tokens_std: torch.LongTensor) -> None:
+    r"""
+    Translates source token probability distributions to target probability distributions, by
+    aligning segments through source token splits, then greedily performing one-to-one,
+    one-to-many, many-to-one distribution mappings.
+        Args:
+            probs (:obj:`torch.FloatTensor`, `required`):
+                [sequence_len, vocab_size] Input probability distribution over a source tokenizer vocabulary.
+            probs_std (:obj:`torch.FloatTensor`, `required`):
+                [std_sequence_len, std_vocab_size] Output probability distribution over a target tokenizer vocabulary.
+                Reference that will be written in-place.
+            offset_mapping (:obj:`List[tuple]`, `required`):
+                Tokenizer offset mappings for a specific sequence [(left_0, right_0), (left_1, right_1), ...].
+            offset_mapping_std (:obj:`List[tuple]`, `required`):
+                Standard tokenizer offset mappings for a specific sequence [(left_0, right_0), (left_1, right_1), ...]
+            tokenizer (:obj:`PreTrainedTokenizerBase`, `required`):
+                Source tokenizer.
+            std_tokenizer (:obj:`PreTrainedTokenizerBase`, `required`):
```

Comment:
Doesn't look like this actually is used in this function

Reply:
Need these parameters for debugging later on, so keeping it for now.